### DISCUSSION TITLE
[WebdriverIO+Appium 2.0] element.waitForExists() doesn't wait for an element if that element is not present when mapping it

### DISCUSSION THREAD

**Discussion by gforepsly** [appium/appium#18047] [General]
I have some weird issues when using mentioned frameworks for IOS automated testing. I want to create some kind of dynamic flows in our tests so I don't need to pause or do anything similar when I need to wait for some element.

Let's say our test starts and the app is installed, and the app triggers "Allow" location buttons for a simulator device to allow using location on the device. That is the first element I want to wait for and click when it's displayed.

I would do something like this:
`const getStartedButton = $('~Allow Once');
 await getStartedButton.waitForExists();`

This fails every time. It seems like whenever the IOS element doesn't exist at the moment it's mapped, it will never wait for that element

I would do something like this, but with wait:
`const getStartedButton = $('~Allow Once');
 await driver.pause(5000);
 await getStartedButton.waitForExists();`

Then this works fine. What am I doing wrong?
It's the same when using WebdriverIO in sync and async mode. 
Currently we are on latest Node 14, Appium beta 2.0 version 48, WebdriverIO on 7.29.1, and our project is a typescript project, se we also have tsconfig with just default stuff from the documentation.

**Comment by jlipps** [comment]
Should probably ask this in the WebdriverIO forums, as it's not really an Appium question. I'm not sure how `waitForExists` is supposed to work.

But pretty sure you also need to `await` the `$` function?

**Reply by gforepsly** [reply]
I will try there as well. It's completely the same if I await for the element and do it like this:
`const getStartedButton = await $('~Allow Once'); await driver.pause(5000); await getStartedButton.waitForExists();`

**Comment by gforepsly** [comment]
I will copy my answer from the other topic, maybe it helps someone as well:

"So, I did a lot of tries and I can say that the "sync" option in WebdriverIO is the cause of the issues. As soon as I installed everything from scratch and started fully with async (also preferred) WebdriverIO way - all of "expect" assertions are working as expected.

The problem is that our project has more than 2 years, so with every WebdriverIO major version update something new was different, which ultimately now means that even with Webdriver 7.x and NodeJs 14 (last versions where sync syntax should work) it doesn't work all they way. We are refactoring our code and that will solve all of our isses.

Thx for the tips - We also check for presence before some actions like click etc.; this example was just for the sake of example."
